util/packages.go

```diff
@@ -19,84 +20,326 @@ import (
-	if packageList != nil {
-		return packageList, nil
+func (i *FileSystemIndexer) GetPackages(ctx context.Context, opts *GetPackagesOptions) (Packages, error) {
+	if i.packageList == nil {
+		var err error
+		i.packageList, err = i.getPackagesFromFileSystem(ctx)
+		if err != nil {
+			return nil, errors.Wrapf(err, "reading packages from filesystem failed")
+		}
 	}
 
-	var err error
-	packageList, err = getPackagesFromFilesystem(ctx, packagesBasePaths)
-	if err != nil {
-		return nil, errors.Wrapf(err, "reading packages from filesystem failed")
+	if opts == nil {
+		return i.packageList, nil
 	}
-	return packageList, nil
+
+	if opts.Filter != nil {
+		return opts.Filter.Apply(ctx, i.packageList), nil
+	}
+
+	return i.packageList, nil
 }
 
-func getPackagesFromFilesystem(ctx context.Context, packagesBasePaths []string) (Packages, error) {
-	span, ctx := apm.StartSpan(ctx, "GetPackagesFromFilesystem", "app")
+func (i *FileSystemIndexer) getPackagesFromFileSystem(ctx context.Context) (Packages, error) {
+	span, ctx := apm.StartSpan(ctx, "GetPackagesFromFileSystem", "app")
+	span.Context.SetLabel("indexer", i.label)
 	defer span.End()
 
-	packagePaths, err := getPackagePaths(packagesBasePaths)
-	if err != nil {
-		return nil, err
+	type packageKey struct {
+		name    string
+		version string
 	}
+	packagesFound := make(map[packageKey]struct{})
 
 	var pList Packages
-	for _, path := range packagePaths {
-		p, err := NewPackage(path)
+	for _, basePath := range i.paths {
+		packagePaths, err := i.getPackagePaths(basePath)
 		if err != nil {
-			return nil, errors.Wrapf(err, "loading package failed (path: %s)", path)
+			return nil, err
 		}
 
-		pList = append(pList, *p)
+		log.Printf("Packages in %s:", basePath)
+		for _, path := range packagePaths {
+			p, err := NewPackage(path, i.fsBuilder)
+			if err != nil {
+				return nil, errors.Wrapf(err, "loading package failed (path: %s)", path)
+			}
+
+			key := packageKey{name: p.Name, version: p.Version}
+			if _, found := packagesFound[key]; found {
+				log.Printf("%-20s\t%10s\t%s", p.Name+" (duplicated)", p.Version, p.BasePath)
+				continue
+			}
+
+			packagesFound[key] = struct{}{}
+			pList = append(pList, p)
+
+			log.Printf("%-20s\t%10s\t%s", p.Name, p.Version, p.BasePath)
+		}
 	}
 	return pList, nil
 }
 
 // getPackagePaths returns list of available packages, one for each version.
-func getPackagePaths(allPaths []string) ([]string, error) {
+func (i *FileSystemIndexer) getPackagePaths(packagesPath string) ([]string, error) {
 	var foundPaths []string
-	for _, packagesPath := range allPaths {
-		log.Printf("Packages in %s:", packagesPath)
-		err := filepath.Walk(packagesPath, func(path string, info os.FileInfo, err error) error {
-			relativePath, err := filepath.Rel(packagesPath, path)
-			if err != nil {
-				return err
-			}
+	err := filepath.Walk(packagesPath, func(path string, info os.FileInfo, err error) error {
```

Comment:
You can also consider WalkDir as it might be a bit more efficient for raw file system.

Reply:
Good idea, done.